includes/class-wc-coupon.php

```diff
@@ -154,131 +322,503 @@ private function get_coupon_id_from_code( $code ) {
-				} elseif ( 'discount_type' === $key ) {
-					$this->discount_type = ! empty( $data[ 'type' ] ) ? wc_clean( $data[ 'type' ] ) : $this->discount_type;
+		$discount = min( $discount, $discounting_amount );
+
+		// Handle the limit_usage_to_x_items option
+		if ( $this->is_type( array( 'percent_product', 'fixed_product' ) ) ) {
+			if ( $discounting_amount ) {
+				if ( '' === $this->get_limit_usage_to_x_items() ) {
+					$limit_usage_qty = $cart_item_qty;
+				} else {
+					$limit_usage_qty = min( $this->get_limit_usage_to_x_items(), $cart_item_qty );
+					$this->set_limit_usage_to_x_items( max( 0, $this->get_limit_usage_to_x_items() - $limit_usage_qty ) );
+				}
+				if ( $single ) {
+					$discount = ( $discount * $limit_usage_qty ) / $cart_item_qty;
+				} else {
+					$discount = ( $discount / $cart_item_qty ) * $limit_usage_qty;
 				}
 			}
+		}
 
-			if ( empty( $this->$key ) ) {
-				$this->$key = $value;
-			} elseif ( in_array( $key, array( 'product_ids', 'exclude_product_ids', 'product_categories', 'exclude_product_categories', 'customer_email' ) ) ) {
-				$this->$key = $this->format_array( $this->$key );
-			} elseif ( in_array( $key, array( 'usage_limit', 'usage_limit_per_user', 'limit_usage_to_x_items', 'usage_count' ) ) ) {
-				$this->$key = absint( $this->$key );
-			} elseif( 'expiry_date' === $key ) {
-				$this->expiry_date = $this->expiry_date && ! is_numeric( $this->expiry_date ) ? strtotime( $this->expiry_date ) : $this->expiry_date;
-			}
+		$discount = round( $discount, WC_ROUNDING_PRECISION );
+
+		return apply_filters( 'woocommerce_coupon_get_discount_amount', $discount, $discounting_amount, $cart_item, $single, $this );
+	}
+
+	/*
+	|--------------------------------------------------------------------------
+	| Setters
+	|--------------------------------------------------------------------------
+	|
+	| Functions for setting coupon data. These should not update anything in the
+	| database itself and should only change what is stored in the class
+	| object.
+	|
+	*/
+
+	/**
+	 * Set coupon code.
+	 * @since  2.6.0
+	 * @param  string $code
+	 */
+	public function set_code( $code ) {
+		$this->_data['code'] = apply_filters( 'woocommerce_coupon_code', $code );
+	}
+
+	/**
+	 * Set coupon description.
+	 * @since  2.6.0
+	 * @param  string $description
+	 */
+	public function set_description( $description ) {
+		$this->_data['description'] = $description;
+	}
+
+	/**
+	 * Set discount type.
+	 * @since  2.6.0
+	 * @param  string $discount_type
+	 */
+	public function set_discount_type( $discount_type ) {
+		$this->_data['discount_type'] = $discount_type;
+	}
+
+	/**
+	 * Set amount.
+	 * @since  2.6.0
+	 * @param  float $amount
+	 */
+	public function set_amount( $amount ) {
+		$this->_data['amount'] = wc_format_decimal( $amount );
+	}
+
+	/**
+	 * Set expiration date.
+	 * @since  2.6.0
+	 * @param  string $date
+	 */
+	public function set_expiry_date( $date ) {
+		$this->_data['expiry_date'] = $date;
```

Comment:
Lets handle these based on if numeric and turn into a timestamp so it's always stored within this class consistently.